app/queries/get_grouped_content_and_links.rb

```diff
@@ -0,0 +1,116 @@
+module Queries
+  class GetGroupedContentAndLinks
```

Comment:
I've added a comment further down that explains why pagination works the way it does.